kill all child processes

jhadvig opened the pull request:
The OS server has to be started without sudo or else the the cleanup wont be able shut the server down, since sudo will spawn a new subshell. When trying to kill the OS server, only the subshell will be killed and the OS server will become a orphan process and then needs to be killed manually.

This fix counts with the case that we are running our tests as sudo/root
@deads2k thoughts ?

deads2k commented on hack/util.sh:
```diff
@@ -96,7 +96,7 @@ function start_os_server {
       1>&2 2>"${BASETMPDIR}/openshift.log" &
   else
-    sudo env "PATH=${PATH}" OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start \
+    env "PATH=${PATH}" OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start \
```
`sudo` is required to be able to make services work correctly, so without privileged execution, the value of starting the server is severely limited.  I thought that @smarterclayton had some bash insanity that got the job done.

smarterclayton replied:
Sudo is required in hack/util.sh for start_os_server for the all-in-one
because you need root.  We don't require running our hack scripts with
sudo.  So we need to continue to use sudo for running a full server.

On Mon, Aug 10, 2015 at 7:49 AM, David Eads <notifications@github.com>
wrote:

> In hack/util.sh
> <https://github.com/openshift/origin/pull/4072#discussion_r36623156>:
>
> > @@ -96,7 +96,7 @@ function start_os_server {
> >        --loglevel=4 \
> >        1>&2 2>"${BASETMPDIR}/openshift.log" &
> >    else
> > -    sudo env "PATH=${PATH}" OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start \
> > +    env "PATH=${PATH}" OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start \
>
> sudo is required to be able to make services work correctly, so without
> privileged execution, the value of starting the server is severely limited.
> I thought that @smarterclayton <https://github.com/smarterclayton> had
> some bash insanity that got the job done.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4072/files#r36623156>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
Our tests should not require root - we should fix the process monitoring though.

jhadvig commented:
@smarterclayton I've tried to different approchas to this issue(killing the whole processs group, ...) but without any luck. The server will be still running after the cleanup since we are running with sudo.
Will revert the changes that had been made with the extended tests to see if the issue had not been introduced there.

smarterclayton commented:
Do we get the sudo subshell pid?  Can we then sudo kill its group?

On Mon, Aug 10, 2015 at 1:36 PM, Jakub Hadvig <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> I've tried to
> different approchas to this issue(killing the whole processs group, ...)
> but without any luck. The server will be still running after the cleanup
> since we are running with sudo.
> Will revert the changes that had been made with the extended tests to see
> if the issue had not been introduced there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4072#issuecomment-129543164>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

jhadvig commented:
@smarterclayton yes we get it. I've tried to kill the whole GROUP but that will exit the script (eg. run.sh) with non-zero return value.

jhadvig commented:
@deads2k I've update the `cleanup()` and should be working now. All child processes of run.sh are killed.
The only other "issue" I've found is that `stop_openshift_server` wont actually stop the the os server cause the OS_PID contains the pid ofthe sudo subshell not the os server process itself. But that not a big deal since all child process are killed by the kill on run.sh#23
In that case I will get rid of the method call. Thoughts ?

deads2k commented:
> In that case I will get rid of the method call. Thoughts ?

No objection to this being a first step.  I do think that we should consider trying to refactor most of `cleanup` to be common, but I won't stop this pull on it.

Can you try making use of the new mechanisms in `test-cmd.sh` to undo @smarterclayton's revert?  If openshift wasn't getting killed properly from `test-end-to-end.sh`, that's probably why `test-extended.sh` was failing on Jenkins.

[test]

jhadvig commented:
@deads2k ok will nuke the call and will have a closer look on the unifying the `cleanup` and undo the revert. But since I don't think it will be done quickly, I would go ahead and merge this, since some guys from our team are working on new ext-tests so they could pull this change. 

> If openshift wasn't getting killed properly from test-end-to-end.sh, that's probably why test-extended.sh was failing on Jenkins.
Will test that.

deads2k commented:
Can you fix `test-cmd.sh` here?  I'm definitely fine with waiting to unify `cleanup`.

jhadvig commented:
@deads2k @gabemontero so I've found out we have bit of a problem here. 
Since we are running the os server start with sudo, all the time we are trying to shut down the os server with $OS_PID(which we are setting in the https://github.com/openshift/origin/blob/master/hack/util.sh#L105) or by obtaining all the running processes of current shell (part of this PR, run.sh#21) the pid we get is not the pid of the os server but actually pid of sudo instead. I thought that by killing the sudo process I would kill the its child process (which is the os server itself), but it wont, it will just kill the sudo process.
This is also issue in the test-end-to-end.sh, here is the output from its `cleanup()`. After the test I'm checking whether the os server is still running http://pastebin.test.redhat.com/304593
So my guess is that either os server is not handling the `kill` (`kill -9` as well) signal appropriately or its waiting for something, since killing the parent sudo process should also kill child os server.

The weird thing is also that then I do `sudo kill -9 os_pid` the os server will be killed immediately. I was also checking the os server logs with loglevel 5, but nothing interesting appeared there http://pastebin.test.redhat.com/304600

I was also asking @mmahut from ops and he was also surprise from this issue.

If you have any thoughts, I would appreciate them.

fyi: from what I was searching there is now way how to obtain child processes of sudo.

gabemontero commented:
@jhadvig here is my 2 cents (where there a clarification seeking questions as much as anything) where I got a little background from @bparees earlier today .... I definitely landed in a KISS sort of space after my initial pondering on this one:

0) assuming there was reluctance or a reason I am missing about using `sudo killall openshift`

1) So, at least when this stuff is running as part of our jenkins job, it has its own ec2 instance, so there is no chance of clobbering something else if you just sudo killall openshift (assuming we can use killall when running as part of our jenkins test run) - no PID maintenance, killall seems sufficient

2) Even with parallel tests, I hear there will still only be 1 openshift server running, Just killall at the end

3) Perhaps accounting for folks running test-extended.sh locally on their system, killing the instance we launch vs. using the instance they have launched prior to the running test-exteneded.sh, perhaps we could gate the use of killall / any killing in general based on whether we reused an already running image, where that var is by default not set, but or an ENV var that is set a certain way in the case of the jenkins jobs, but if the developer explicitly sets it, or if we don't see openshift running locally when the tests start, we just killall

Is the above rationale not sufficient and/or what is it overlooking ?

bparees commented:
@jhadvig I pretty much agree with @gabemontero that killlall is probably ok here, but if you want to get fancy (and i know you do!), you could do this:

1) grab the sudo pid when you run the sudo command
2) the sudo pid is also the group id (pgid), so then you can do:
ps ax o pid,pgid | grep $SUDO_PID | awk '{print $1}' | xargs sudo kill -9

smarterclayton commented:
I think bash is officially on my dead to me list.

On Aug 12, 2015, at 4:17 PM, Ben Parees <notifications@github.com> wrote:

@jhadvig <https://github.com/jhadvig> I pretty much agree with @gabemontero
<https://github.com/gabemontero> that killlall is probably ok here, but if
you want to get fancy (and i know you do!), you could do this:

1) grab the sudo pid when you run the sudo command
2) the sudo pid is also the group id (pgid), so then you can do:
ps ax o pid,pgid | grep $SUDO_PID | awk '{print $1}' | xargs sudo kill -9

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4072#issuecomment-130432776>.

bparees commented:
@smarterclayton well whatever you do, don't look over here then:
https://github.com/openshift/jenkins/blob/master/1.6/contrib/configuration/jobs/OpenShift%20Sample/config.xml#L158-L162

(the escaping makes it look slightly worse than it is...)

smarterclayton commented:
Hsssssssssssssssssssss

On Aug 12, 2015, at 11:09 PM, Ben Parees <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> well whatever you do,
don't look over here then:
https://github.com/openshift/jenkins/blob/master/1.6/contrib/configuration/jobs/OpenShift%20Sample/config.xml#L158-L162

(the escaping makes it look slightly worse than it is...)

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4072#issuecomment-130514661>.

jhadvig commented:
@gabemontero well thats a path I didnt want to take. Since we are not running the tests are non-root I guess we assume that somebody will also run the tests locally not just in a sandBox. Therefor the `killall` solution will kill everything (vim with opened *openshift* file, sshd to *openshift*, and god knows what else). Thats why I would like to come with more sophisticated way.

@bparees the problem is not that I cant kill the sudo process and it will die, but the problem that the os server wont. Don't understand why but it stay running (tried kill and kill -9). After the scripts finishes and I try to kill the server manually it will die.

deads2k commented:
This might have been discussed before, but why not
```
pids="$(jobs -pr)"
# loop over pids
    ps --ppid=${pids} | xargs sudo kill 
    sudo kill ${pids}
```

I'll leave the loop to you, but that ought to kill all the processes created by all the jobs created by this bash script, right?

gabemontero commented:
@jhadvig sure thing, as I mentioned before, the `killall` is the KISS / meets min sort of thing.  But it looks like that between you and @deads2k the more sophisticated solution is close.

bparees commented:
@jhadvig you're saying that a direct kill -9 on the openshift process pid (not the sudo pid) is not killing openshift?

bparees commented:
and @deads2k your proposal doesn't seem to catch the process launched by sudo:
$ sudo sleep 100

$ ps ax | grep sleep
 2698 pts/2    S+     0:00 sudo sleep 100
 2703 pts/2    S+     0:00 sleep 100

$ ps --pid=2698
  PID TTY          TIME CMD
 2698 pts/2    00:00:00 sudo

jhadvig commented:
@bparees I dont know why but no upon the `cleanup()`. But the approach @deads2k suggested is wroking, at least it seems so.

deads2k commented:
> and @deads2k your proposal doesn't seem to catch the process launched by sudo:

@bparees that extra `p` in `ppid` isn't a typo :)

bparees commented:
@deads2k i knew that but it was a typo when i left it out while trying it :)

deads2k commented:
> Can you fix test-cmd.sh here? It seems like you fixed at least part of the problem he was having. I'm definitely fine with waiting to unify cleanup.

I'd still like to see this here to prove that fix works correctly in all cases.

jhadvig commented:
@deads2k regarding the `test-cmd`, I would rather leave it as it is. The script is working and also is clearing all the processes at the end. The thing is that we launching `test-cmd` also in Travis where we dont have sudo and other things(cant bind to 0.0.0.0). thats why the setup and cleanup process will look differently from other tests (`test-*.sh`). For that reason I think that writing a complex but unified setup and cleanup function wont be the happiest way to go with.
Since the `test-cmd` is different from other tests I would leave the logic separate.

I've tested the change and now all the processes spawned by the test-extended/e2e are killed. For the test-cmd the teardown works as well.

deads2k commented:
Is there a reason to not try `kill_all_processes` inside of `test-cmd.sh`?  it seem like `kill_all_processes` could simply test whether sudo is present and call the appropriate kill.

jhadvig commented:
@deads2k well I thought that I would cause problem in the travis since there is no sudo, but I've added a test for the `sudo` and added to the added the `kill_all_processes` to the `test-cmd.sh`.
Thanks for suggestion :+1:

deads2k commented on hack/util.sh:
```diff
@@ -378,20 +353,17 @@ remove_tmp_dir() {
-    fi
-    set -u
-    set -e
+  if type sudo &> /dev/null; then
```
```
sudo=
if type sudo &> /dev/null; then
    sudo=sudo
fi

pids=($(jobs -pr))
for i in ${pids[@]}; do
      ps --ppid=${i} | xargs $sudo kill &> /dev/null
      $sudo kill ${i} &> /dev/null &> /dev/null
done
```

ought to get the job done for either, right?

jhadvig replied:
Yes, this is probably more suitable way.

deads2k commented:
I really want to see travis pass so that @smarterclayton doesn't revert again.

@bparees can you kick travis?

ramr commented:
Not sure about your use case, but I think you can use svr4/posix semantics, just set the session id for the for the command you start and just kill all the processes within that session.  ```man [-a] setsid```
This should catch all the processes created by that ancestor including orphans (that init adopts) unless there is another session leader created in that chain.

For example:  assuming the script is called parent-cmd.sh, you can run it  as  ```setsid sudo parent-cmd.sh```  or even ```sudo setsid parent-cmd.sh```  -- the latter has slightly different semantics. The first case, the session leader is the sudo command and in the latter case its the parent-cmd.sh  

And you can then kill all the processes within a session via something like  
```kill -9 $(ps -s $sessionleader -o pid=)```  where ```$sessionleader``` is the process id of the leader (sudo or parent-cmd.sh in the above example).  HTH

bparees commented:
kicked.

jhadvig commented:
Force pushed, to re-kick the tests.

jhadvig commented:
@deads2k so the Travis is passing, but one of the integration tests failed, but I guess that randomly.

deads2k commented:
one travis success.  [merge]
